### ISSUE TITLE
Feature Request: Allow Customized XDS Address to be set 

### ISSUE THREAD

**Issue by Dr-Octavius** [istio/ztunnel#1203]
## First Issue Intro!
I am an entrepreneur starting out [my software company](https://github.com/sefire-octopi) 
I am going down the kubernetes road for my back-end architecture.
I intend to use Istio as my service mesh in production.
Currently deciding between conventional sidecar or Istio's newer Ambient Mode.

## **Disclaimers**
Issue raised here may be lacking in understanding Istio internals
I understand that Istio Ambient Mode is in Beta & not advised for production.
Configurations discussed are for deploying Istio Ambient Mode to a development environment.
Certain configurations have been sanitized for brevity & security
Referencing #752 for similar issue.

## Technologies
1. Terraform: ```1.9.1``` (HCP Terraform)
2. Helm Terraform Provider: ``2.14.0``
3. Kubernetes Terraform Provider: "2.31.0"
4. Digital Ocean Kubernetes Service: ```1.30.2-do.0``` (Essentially Kubernetes: ```1.30.2``` under the hood)

##  Synopsis & Error Faced
Recently, I tried deploying the Istio Service Mesh on Ambient Mode. 
I did so via Terraform as my CICD pipelines are all written in Terraform Configurations.
I faced the following errors as a result
```bash
2024-07-10T17:00:55.776838Z     info    app     in-pod mode enabled
2024-07-10T17:00:55.776966Z     info    hyper_util      listener established    address=127.0.0.1:15000 component="admin"
2024-07-10T17:00:55.777012Z     info    hyper_util      listener established    address=[::]:15020 component="stats"
2024-07-10T17:00:55.777019Z     info    readiness       Task 'proxy' complete (2.262921ms), still awaiting 2 tasks
2024-07-10T17:00:55.778185Z     info    hyper_util      listener established    address=[::]:15021 component="readiness"
2024-07-10T17:00:55.782698Z     warn    xds::client:xds{id=1}   XDS client connection error: gRPC connection error:status: Unknown, message: "client error (Connect)", source: dns error: failed to lookup address information: Name or service not known, retrying in 20ms
2024-07-10T17:00:55.806177Z     warn    xds::client:xds{id=2}   XDS client connection error: gRPC connection error:status: Unknown, message: "client error (Connect)", source: dns error: failed to lookup address information: Name or service not known, retrying in 40ms
2024-07-10T17:00:55.851253Z     warn    xds::client:xds{id=3}   XDS client connection error: gRPC connection error:status: Unknown, message: "client error (Connect)", source: dns error: failed to lookup address information: Name or service not known, retrying in 80ms
2024-07-10T17:00:55.935488Z     warn    xds::client:xds{id=4}   XDS client connection error: gRPC connection error:status: Unknown, message: "client error (Connect)", source: dns error: failed to lookup address information: Name or service not known, retrying in 160ms
2024-07-10T17:00:56.099556Z     warn    xds::client:xds{id=5}   XDS client connection error: gRPC connection error:status: Unknown, message: "client error (Connect)", source: dns error: failed to lookup address information: Name or service not known, retrying in 320ms
2024-07-10T17:00:56.422465Z     warn    xds::client:xds{id=6}   XDS client connection error: gRPC connection error:status: Unknown, message: "client error (Connect)", source: dns error: failed to lookup address information: Name or service not known, retrying in 640ms
.
.
.
```
The same XDS connection issue will persist

## Feature or Bug?
If Istio's Ambient Mode demands strict standard naming of `istiod`'s CA & XDS Addresse, just 3 questions,
1. Is dynamic naming of the addresses already supported by Istio's installation via Helm's CLI Tool?
2. If not, will dynamic naming of the addresses be supported in Istio's product road map?
3. If no, why or why not?

## Key Information
- Istio Version: ```1.22.2```
- Operating system: ```k8s```

###  Step 1. The Terraform Configuration
I first started out by modifying the Terraform Configuration the best way I knew from [Istio's Installation Guide Using Helm](https://istio.io/latest/docs/ambient/install/helm-installation/)
```tf
#-------------------------------
# Terraform Cloud Configuration
#-------------------------------
terraform {
  cloud {
    organization = "my-org"
    workspaces {
      name = "my-workspace"
    }
  }
  required_providers {
    helm = {
      source  = "hashicorp/helm"
      version = "2.14.0"
    }
    kubernetes = {
      source  = "hashicorp/kubernetes"
      version = "2.31.0"
    }
  }
}
#------------------------
# Provider Configuration
#------------------------
provider "helm" {
  kubernetes {
    config_path = "${path.module}/kubeconfig.yaml"
  }
}
provider "kubernetes" {
  config_path = "${path.module}/kubeconfig.yaml"
}
#-----------------------------------
# Istio k8s Namespace Configuration
#-----------------------------------
resource "kubernetes_namespace" "my_istio_namespace_dev" {
  metadata {
    name = "my-istio-namespace-dev"
  }
}
###################################################
#---------REQUIRED COMPONENTS FOR ISTIO-----------#
# Istio Service Mesh Core Installation using Helm #
#-------------------------------------------------#
###################################################
#------------------------------------
# Istio Base Installation using Helm
#------------------------------------
resource "helm_release" "my_istio_base_name" {
  name             = "my-istio-base-name"
  repository       = "https://istio-release.storage.googleapis.com/charts"
  chart            = "base"
  version          = "1.22.2"
  namespace        = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-base-nodepool"
  }
}
#------------------------------------------------------------------
# Istio CNI Plugin Installation using Helm (For Ambient Mode Beta)
#------------------------------------------------------------------
resource "helm_release" "my_istio_cni_name" {
  name       = "my-istio-cni-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "cni"
  version          = "1.22.2"
  namespace        = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-cni-nodepool"
  }
  set {
    name  = "profile"
    value = "ambient"
  }
  depends_on = [helm_release.my_istio_base_name]
}

#---------------------------------------------------------------
# Istio Discovery (Istio Control Plane) Installation using Helm
#---------------------------------------------------------------
resource "helm_release" "my_istiod_name" {
  name       = "my-istiod-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "istiod"
  version    = "1.22.2"
  namespace  = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "pilot.enabled"
    value = "true"
  }
  set {
    name  = "pilot.autoscaleEnabled"
    value = "true"
  }
  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istiod-nodepool"
  }
  set {
    name  = "profile"
    value = "ambient"
  }
  depends_on = [helm_release.my_istio_cni_name]
}
#---------------------------------------------------------------
# Istio zTunnel Installation using Helm (For Ambient Mode Beta)
#---------------------------------------------------------------
resource "helm_release" "my_istio_ztunnel_name" {
  name       = "my-istio-ztunnel-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "ztunnel"
  version          = "1.22.2"
  namespace        = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-ztunnel-nodepool"
  }
  depends_on = [helm_release.my_istiod_name]
}

#----------------------------------------
# Metrics Server Installation using Helm
#----------------------------------------
resource "helm_release" "my_metrics_server_name" {
  name       = "my-metrics-server-name"
  repository = "https://kubernetes-sigs.github.io/metrics-server/"
  chart      = "metrics-server"
  version    = "3.8.3"
  namespace  = "kube-system"
  set {
    name  = "args[0]"
    value = "--kubelet-insecure-tls"
  }
  set {
    name  = "args[1]"
    value = "--metric-resolution=30s"
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-metrics-server-nodepool"
  }
}
#################################################################
#---------------------------------------------------------------#
# Istio Service Mesh Networking Control Using Helm & Kubernetes #
#---------------------------------------------------------------#
#################################################################
#----------------------------------------------------------------
# Istio k8s Service Mesh Ingress Gateway Installation using Helm
# - INBOUND Traffic Management, Control & Entry Point
#----------------------------------------------------------------
resource "helm_release" "my_istio_ingress_name" {
  name       = "my-istio-ingress-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "gateway"
  version    = "1.22.2"
  namespace  = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "service.type"
    value = "LoadBalancer"
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-ingress-nodepool"
  }
  depends_on = [helm_release.my_istiod_name]
}
#---------------------------------------------------------------
# Istio k8s Service Mesh Egress Gateway Installation using Helm
# - OUTBOUND Traffic Management, Control & Entry Point
#---------------------------------------------------------------
resource "helm_release" "my_istio_egress_name" {
  name       = "my-istio-egress-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "gateway"
  version    = "1.22.2"
  namespace  = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "service.type"
    value = "LoadBalancer"
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-egress-nodepool"
  }
  depends_on = [helm_release.my_istiod_name]
}
```
TLDR; 
The Terraform Configuration File above does the installation sequentially as per [Istio's Installation Guide Using Helm](https://istio.io/latest/docs/ambient/install/helm-installation/) in the following order,  with an additional first step 
1. Create a `kubernetes namespace` called `my-istio-namespace-dev`
3. Install `istio-base` to `my-istio-namespace-dev`
4. Install `istio-cni` to `my-istio-namespace-dev`
5. Install `istiod` to `my-istio-namespace-dev`
6. Concurrently,
    - Install `istio-ingress` to `my-istio-namespace-dev`
    - Install `istio-egress` to `my-istio-namespace-dev`
    - Install `istio-ztunnel` to `my-istio-namespace-dev`
###  Step 2. Error Message
```bash
.
.
.
helm_release.lome_istio_ztunnel_sgp1_dev: Still creating... [5m0s elapsed]
╷
│ Warning: Helm release "my-istio-ztunnel-dev" was created but has a failed status. Use the `helm` command to investigate the error, correct it, then run Terraform again.
│ 
│   with helm_release.my_istio_ztunnel_dev,
│   on main.tf line 157, in resource "helm_release" "my_istio_ztunnel_dev":
│  157: resource "helm_release" "my_istio_ztunnel_dev" {
│ 
╵
╷
│ Error: context deadline exceeded
│ 
│   with helm_release.my_istio_ztunnel_dev,
│   on main.tf line 157, in resource "helm_release" "my_istio_ztunnel_dev":
│  157: resource "helm_release" "my_istio_ztunnel_dev" {
│ 
╵
Operation failed: failed running terraform apply (exit 1)
```
This was the log output of the pipeline. 
Only one issue was observed here. 
All other Istio components were successfully installed
###  Step 3. Error Diagnostics
I downloaded my `kubeconfig.yaml`, set the `PATH` for `kubectl` and ran
```bash
kubectl get events -n my-istio-namespace-dev | findstr ztunnel
```
yielding the following log output,
```bash
9m53s       Normal    Scheduled                      pod/ztunnel-d4vpj                                    Successfully assigned my-istio-namespace-dev/ztunnel-d4vpj to my-istio-ztunnel-nodepool
9m53s       Normal    Pulled                         pod/ztunnel-d4vpj                                     Container image "docker.io/istio/ztunnel:1.22.2" already present on machine
9m53s       Normal    Created                        pod/ztunnel-d4vpj                                     Created container istio-proxy
9m53s       Normal    Started                        pod/ztunnel-d4vpj                                     Started container istio-proxy
4m43s       Warning   Unhealthy                      pod/ztunnel-d4vpj                                     Readiness probe failed: HTTP probe failed with statuscode: 500
```
I proceeded to run,
```bash
kubectl logs ztunnel-d4vpj -n my-istio-namespace-dev
```
yielding me I very long log (I have saved the log somewhere so if anyone on the Istio Team wants to view it let me know!), of which one part stood out
```bash
.
.
.
caAddress: https://istiod.istio-system.svc:XXXXX
caRootCert: !File ./some/secret/path/root-cert.pem
xdsAddress: https://istiod.istio-system.svc:XXXXX
xdsRootCert: !File ./some/secret/path/root-cert.pem
.
.
.
```
I believe the right address should follow a syntax like `<istiod-service-name>.<istio-system-namespace>`. 
This was odd to me since I have already set the namespace explicitly in my Terraform Configuration.
Expected address: `istiod.my-istio-namespace-dev`
###  Step 4. Remediation 1
I attempted to force this change by correcting the zTunnel configuration within my Terraform configuration as such
```tf
#---------------------------------------------------------------
# Istio zTunnel Installation using Helm (For Ambient Mode Beta)
#---------------------------------------------------------------
resource "helm_release" "my_istio_ztunnel_name" {
  name       = "my-istio-ztunnel-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "ztunnel"
  version          = "1.22.2"
  namespace        = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-ztunnel-nodepool"
  }
  #---------------------
  # Added the following
  #---------------------
  set {
    name  = "caAddress"
    value = "https://istiod.${kubernetes_namespace.lome_istio_sgp1_dev_namespace.metadata[0].name}.svc:15012"
  }
  set {
    name  = "xdsAddress"
    value = "https://istiod.${kubernetes_namespace.lome_istio_sgp1_dev_namespace.metadata[0].name}.svc:15012"
  }
  depends_on = [helm_release.my_istiod_name]
}
```
With this configuration, I faced the same error output from the pipeline. Repeating diagnostics steps in Step 3 yielded me the following output,
```bash
.
.
.
caAddress: https://istiod.my-istio-namespace.svc:XXXXX
caRootCert: !File ./some/secret/path/root-cert.pem
xdsAddress: https://istiod.istio-system.svc:XXXXX
xdsRootCert: !File ./some/secret/path/root-cert.pem
.
.
.
```
CA Address changed, but XDS Address remain unchanged.
###  Step 5. Remediation 2
Perhaps a syntax change. 
I tried again to force this by correcting the zTunnel configuration within my Terraform configuration as such,
```tf
#---------------------------------------------------------------
# Istio zTunnel Installation using Helm (For Ambient Mode Beta)
#---------------------------------------------------------------
resource "helm_release" "my_istio_ztunnel_name" {
  name       = "my-istio-ztunnel-name"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "ztunnel"
  version          = "1.22.2"
  namespace        = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace_dev.metadata.0.name
  }
  set {
    name  = "global.nodeSelector.nodepool"
    value = "my-istio-ztunnel-nodepool"
  }
  set {
    name  = "caAddress"
    value = "https://istiod.${kubernetes_namespace.lome_istio_sgp1_dev_namespace.metadata[0].name}.svc:15012"
  }
  #-----------------------
  # Amended the following
  #-----------------------
  set {
    name  = "xds.address"
    value = "https://istiod.${kubernetes_namespace.lome_istio_sgp1_dev_namespace.metadata[0].name}.svc:15012"
  }
  depends_on = [helm_release.my_istiod_name]
}
```
With this configuration, I faced exactly the same error output from the pipeline, with the same diagnostic output in Step 4.
CA Address changed, but XDS Address remain unchanged.
###  Step 5. Remediation 3
Finally, I decided to slowly amend naming based on [Istio's Installation Guide Using Helm](https://istio.io/latest/docs/ambient/install/helm-installation/)
```tf
#-----------------------------------
# Istio k8s Namespace Configuration
#-----------------------------------
resource "kubernetes_namespace" "my_istio_namespace_dev" {
  #-----------------------
  # Amended the following
  #-----------------------
  metadata {
    name = "istio-system"
  }
}
```
And the zTunnel worked!

##  Final Thoughts
I guess, I just want to know if it is possible (if anyone in the community has been able) to set the XDS & CA Addresses dynamically based on Terraform or CLI, or if I am just not using helm properly.

Apologies for the Long Read 

EDIT:
## Why this matters?
- It would be good to not rely on a single `istio-system` naming as it becomes difficult to identify between development, release, staging, production etc. environments especially when doing things in an Infrastructure-As-Code (IaC) level
- Whilst places like [artifacthub](https://artifacthub.io/) contains helm charts for easy reference to use Terraform to deploy code, documentation on the XDR Address is not clear (or perhaps I am just a novice LOL)

**Comment by howardjohn** [comment]
Hey! Thanks for the issue.

I think your config isn't working for 2 reasons, both of which are recently fixed (or in progress), but not yet landed:

1. You set global.istioNamespace. This doesn't work yet in Helm (but would with istioctl), but should with https://github.com/istio/istio/pull/51749/. You can workaround this today in 1.22 with `istioNamespace` (not under global)
2. `xdsAddress` would also work, but this setting was added after 1.22, so it has no impact in 1.22

So I would recommend for now you just set `istioNamespace` instead of `global.IistioNamespace` and I think that should unblock you

**Comment by Dr-Octavius** [comment]
Thank you @howardjohn for the reply and the reference to [istio#517](https://github.com/istio/istio/pull/51749). Things are much clearer now.

I see this is an ongoing PR. 
Are there any other fields within global where this behavior is expected?
Or is it recommended to skip the use of `global` in general for 1.22?

**Comment by howardjohn** [comment]
Ztunnel chart has no global in 1.22, only in 1.23. So for 1.22 anything
under global will do nothing

**Comment by Dr-Octavius** [comment]
@howardjohn that's great advice!! Thank you so much

**Comment by Dr-Octavius** [comment]
Hey @howardjohn I wanted to post some updates and officially close this.

Your suggestions helped greatly. This was the final configuration I ended up with
```tf
#---------------------------------------------------------------
# Istio zTunnel Installation using Helm (For Ambient Mode Beta)
#---------------------------------------------------------------
resource "helm_release" "my_istio_ztunnel" {
  name       = "my-istio-ztunnel"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "ztunnel"
  version          = "1.22.2"
  namespace        = kubernetes_namespace.my_istio_namespace.metadata.0.name
  
  set {
    #------------------------DO NOT AMEND-----------------------
    # For 1.22, should not change this to global.istioNamespace
    #-----------------------------------------------------------
    name  = "istioNamespace"
    value = kubernetes_namespace.my_istio_namespace.metadata.0.name
  }

  set {
    name  = "nodeSelector.nodepool"
    value = "my-zTunnel-nodepool"
  }
  
  #---------------------------------
  # Enabled For Multi-Cluster Istio
  #---------------------------------
  set {
    name  = "multiCluster.clusterName"
    value = "my-istio-cluster"
  }
  
  depends_on = [
    helm_release.my_istio_ingress,
    helm_release.my_istio_egress
  ]
}
```
Additionally, I even managed to spin up the multi-cluster configuration as well! Kudos to the Istio Team for the hard work!
Below is the accompanying `istiod` using Terraform's `helm_release` for others to reference
```tf
#---------------------------------------------------------------
# Istio Discovery (Istio Control Plane) Installation using Helm
#---------------------------------------------------------------
resource "helm_release" "my_istiod" {
  name       = "my-istiod"
  repository = "https://istio-release.storage.googleapis.com/charts"
  chart      = "istiod"
  version    = "1.22.2"
  namespace  = kubernetes_namespace.my_istio_namespace.metadata.0.name

  set {
    name  = "profile"
    value = "ambient"
  }

  set {
    name  = "pilot.enabled"
    value = "true"
  }

  set {
    name  = "pilot.autoscaleEnabled"
    value = "true"
  }

  set {
    name  = "pilot.autoscaleMin"
    value = "1"
  }

  set {
    name  = "pilot.autoscaleMax"
    value = "2"
  }

  set {
    name  = "pilot.nodeSelector.nodepool"
    value = "my-istio-d-nodepool"
  }

  set {
    name  = "global.istioNamespace"
    value = kubernetes_namespace.my_istio_namespace.metadata.0.name
  }
  
  #---------------------------------
  # Enabled For Multi-Cluster Istio
  #---------------------------------
  set {
    name  = "global.multiCluster.enabled"
    value = "true"
  }

  set {
    name  = "global.multiCluster.clusterName"
    value = "my-istio-cluster"
  }

  depends_on = [helm_release.my_istio_cni]
}
```
Importantly, I managed to reconcile your suggestions with the Helm Chart and realized I was referencing it wrong.

**Comment by howardjohn** [comment]
Awesome, glad to hear you got it working

**Comment by howardjohn** [comment]
I am going to close this out since its resolved, let us know if you run into other issues. Thanks!
